### PR TITLE
bug fix and implementation on `denormalization` logic.

### PR DESCRIPTION
**Pull request by jinminhao**
- [x] bug fix on merge_attr
- [x] bug fix on generate and write to csv
- [x] bug fix on denormalziation for continuous field when activating the logp flag 
- [x] add a new field called Word2Vec.

### CODE REVIEW

**Comment by jinminhao** [comment]
I would try to finish the work and close this pr before 1/19. Thanks!

File: netshare/generate/model_generate.py [hunk 1]
```diff
@@ -72,19 +73,25 @@ def generate_data(
...
+    :return: None
     """
     chunk0_idx = config_group["config_ids"][0]
     eval_root_folder = configs[chunk0_idx]["eval_root_folder"]
     attr_raw_npz_folder = os.path.join(eval_root_folder, "attr_raw")
-    word2vec_size = configs[chunk0_idx].get("word2vec_vecSize", 0)
-    pcap_interarrival = configs[chunk0_idx].get("timestamp") == "interarrival"
     num_chunks = len(config_group["config_ids"])
 
-    if not pcap_interarrival:
```
Comment (jinminhao) [inline netshare/generate/model_generate.py:84]: The logic here is to find the column index which is a *flag* used to represent if the session (flow) start in this chunk.

Previously, the attribute schema is (5-tuple, [session_start_time], *flag*, ....).  Therefore, we need to have a if condition to get the correct index based on if the user specifies `interarrival`(`interarrival` means there is a session_start_time in attributes).

But, after the code refactoring, the schema is (5-tuple, *flag*, ...., , [session_start_time]). Therefore, we don't need this if condition logic anymore.

File: netshare/generate/model_generate.py [hunk 2]
```diff
@@ -112,44 +119,28 @@ def merge_attr(config_group: dict, configs: List[dict]) -> None:
             os.path.join(attr_raw_npz_folder, "chunk_id-{}.npz".format(chunkid))
         )["data_attribute"]
 
-        if num_chunks > 1 and word2vec_size:
```
Comment (jinminhao) [inline netshare/generate/model_generate.py:115]: Whether we need to concat and merge the attributes is not based on if the user specifies `word2vec_size`. I just deleted that condition here.

File: netshare/generate/model_generate.py [hunk 3]
```diff
@@ -112,44 +119,28 @@ def merge_attr(config_group: dict, configs: List[dict]) -> None:
...
-                    row_this_chunk += [1.0, 0.0] * (chunkid + 1)
-                    for i in range(chunkid + 1, num_chunks):
-                        if (
-                            row[bit_idx_flagstart + 2 * i + 2]
-                            < row[bit_idx_flagstart + 2 * i + 3]
-                        ):
-                            row_this_chunk += [0.0, 1.0]
-                        else:
-                            row_this_chunk += [1.0, 0.0]
-                    # dict_chunkid_attr[chunkid].append(row_this_chunk)
```
Comment (jinminhao) [inline netshare/generate/model_generate.py:135]: the `row_this_chunk` is constructed but not used in the following logic. I just deleted them.

File: netshare/generate/model_generate.py [hunk 4]
```diff
@@ -160,15 +151,3 @@ def merge_attr(config_group: dict, configs: List[dict]) -> None:
             os.path.join(attr_clean_npz_folder, "chunk_id-{}.npz".format(chunkid)),
             data_attribute=np.asarray(attr_clean),
         )
-
-
-# ===================== TODO: move merge_syn_df to postprocess ================
-
-
-def get_per_chunk_df(chunk_folder: str) -> pd.DataFrame:
-    '''chunk_folder: "chunk_id-0"'''
```
Comment (jinminhao) [inline netshare/generate/model_generate.py:169]: This function is no longer in use I guess.

**Comment by sniperyyc** [comment]
> I would try to finish the work and close this pr before 1/19. Thanks!

No rush, please take your time :)

File: netshare/generate/denormalize_fields.py [hunk 5]
```diff
@@ -32,14 +38,62 @@ def _denormalize_by_fields_list(
...
         if is_session_key:
-            sub_data = normalized_data[:, dim : dim + field.getOutputDim()]
+            sub_data = normalized_data[:, dim: dim + field.getOutputDim()]
         else:
-            sub_data = normalized_data[:, :, dim : dim + field.getOutputDim()]
+            sub_data = normalized_data[:, :, dim: dim + field.getOutputDim()]
+
         sub_data = field.denormalize(sub_data)
-        if isinstance(field, ContinuousField):
-            sub_data = sub_data[:, 0] if is_session_key else sub_data[:, :, 0]
```
Comment (jinminhao) [inline netshare/generate/denormalize_fields.py:42]: This may be incorrect. We cannot only get the first dimension on the continuous fields when denormalizing.

File: netshare/generate/model_generate.py (hunk 1 above)
Comment (saart) [reply netshare/generate/model_generate.py:84]: Oh I'm surprised that I changed the schema. 
I'm so sorry for not noticing that!
I hope that the stronger E2E tests that Dorothy currently writing will help make sure that it won't happen again!
Thank you!!

File: netshare/generate/model_generate.py [hunk 6]
```diff
@@ -72,19 +73,25 @@ def generate_data(
...
+    chunks of data will first generate attributes. Then, this function will be
+    invoked to merge the synthetic attributes. Given the merged attributes, models
+    will continue generating the time-series measurements.
+
+    :return: None
     """
     chunk0_idx = config_group["config_ids"][0]
     eval_root_folder = configs[chunk0_idx]["eval_root_folder"]
     attr_raw_npz_folder = os.path.join(eval_root_folder, "attr_raw")
-    word2vec_size = configs[chunk0_idx].get("word2vec_vecSize", 0)
```
Comment (saart) [inline netshare/generate/model_generate.py]: I think that no one is using the `word2vec_vecSize` config now, right?
We can drop it 🎉

File: netshare/generate/model_generate.py [hunk 7]
```diff
@@ -72,19 +73,25 @@ def generate_data(
...
-        bit_idx_flagstart = 128 + word2vec_size * 3
-    else:
-        bit_idx_flagstart = 128 + word2vec_size * 3 + 1
+    dim = 0
+    for field in list(learn_api.get_attributes_fields().values()):
+        if field.name != "startFromThisChunk":
+            dim += field.getOutputDim()
+        else:
+            break
+    col_idx_of_session_start = dim
```
Comment (saart) [inline netshare/generate/model_generate.py:94]: your logic depends on the order of the list, and the list is stemmed from a `dict`'s values.
Generally, we shouldn't do that because `dict` is an unordered data structure.
Maybe we should use an OrderedDict here? Alternatively, I'm really not sure why the API stores `dict` and not `list`. We need the dictionary due to the double parsing in `setup_cross_chunks_data:155` and `setup_per_chunk_data:44` (which is really bad and should be removed as well!!), so maybe we can store `list` in the API? I.e. change `learn_api:102`?

File: netshare/generate/model_generate.py [hunk 8]
```diff
@@ -72,19 +73,25 @@ def generate_data(
...
-    word2vec_size = configs[chunk0_idx].get("word2vec_vecSize", 0)
-    pcap_interarrival = configs[chunk0_idx].get("timestamp") == "interarrival"
     num_chunks = len(config_group["config_ids"])
 
-    if not pcap_interarrival:
-        bit_idx_flagstart = 128 + word2vec_size * 3
-    else:
-        bit_idx_flagstart = 128 + word2vec_size * 3 + 1
+    dim = 0
+    for field in list(learn_api.get_attributes_fields().values()):
```
Comment (saart) [inline netshare/generate/model_generate.py:89]:
```suggestion
    for field in learn_api.get_attributes_fields().values():
```
This is more pythonic (and better performance) :)

File: netshare/learn/utils/word2vec_embedding.py [hunk 9]
```diff
@@ -36,3 +41,90 @@ def word2vec_train(
...
+        type = col.encoding.split("_")[1]
+        if type not in dict_type_cols:
+            dict_type_cols[type] = []
+        dict_type_cols[type].append(col.column)
+    return dict_type_cols
+
+
+def build_annoy_dictionary_word2vec(
+    df, model_path, word2vec_cols, word2vec_size, n_trees
+):
```
Comment (saart) [inline netshare/learn/utils/word2vec_embedding.py]: Can you use types throughout the file?

File: netshare/learn/utils/word2vec_embedding.py [hunk 10]
```diff
@@ -36,3 +41,90 @@ def word2vec_train(
...
+
+    model = Word2Vec.load(model_path)
+    wv = model.wv
+
+    # type : [cols]
+    # ("ip": ["srcip", "dstip"])
+    # "port": ["srcport", "dstport"]
+    # "proto": ["proto"]
+    dict_type_cols = get_word2vec_type_col(word2vec_cols)
+    logger.info(dict_type_cols)
```
Comment (saart) [inline netshare/learn/utils/word2vec_embedding.py]: please write an informative log, and it should probably be a debug

File: netshare/generate/denormalize_fields.py [hunk 11]
```diff
@@ -32,14 +41,66 @@ def _denormalize_by_fields_list(
...
+        for m in (session_key_fields + word2vec_config.get("timeseries", []))
+        if "word2vec" in m.get("encoding", "")
+    ]
+    word2vec_model_path = os.path.join(
+        learn_api.get_word2vec_model_directory(),
+        "{}_{}.model".format(
+            word2vec_config["word2vec"]["model_name"],
+            word2vec_config["word2vec"]["vec_size"],
+        ),
+    )
```
Comment (saart) [inline netshare/generate/denormalize_fields.py]: If that's possible, can you put these lines as a new API in the `learn_api` module, e.g. something like `get_word2vec_model_path`, and make sure that the writing procedure (`word2vec_train`) also uses it?

File: netshare/generate/denormalize_fields.py [hunk 12]
```diff
@@ -32,14 +41,66 @@ def _denormalize_by_fields_list(
...
+    word2vec_cols = [
+        m
+        for m in (session_key_fields + word2vec_config.get("timeseries", []))
+        if "word2vec" in m.get("encoding", "")
+    ]
+    word2vec_cols_names = [
+        m["column"]
+        for m in (session_key_fields + word2vec_config.get("timeseries", []))
+        if "word2vec" in m.get("encoding", "")
+    ]
```
Comment (saart) [inline netshare/generate/denormalize_fields.py]: in all these lines you're reading and parsing the learning configuration. It's a code smell that the parsing logic happens twice, in two different components.
Here's a suggestion that will make our life much much easier:
Add the member `is_word2vec_encoding` to `Field` with the default value `False`. 
Populate it during the instance creation in the module `setup_cross_chunks_data` line 163. 
then, here, you can just filter the given `fields_list` using this member.
I.e. this is better because of two big reasons:
1. We encapsulate the configuration parsing logic in a single place
2. you replace 16 lines of complex code (the is a duplication in some sense) with 4 lines of easy code :-)

File: netshare/generate/denormalize_fields.py [hunk 13]
```diff
@@ -32,14 +41,66 @@ def _denormalize_by_fields_list(
...
             sub_data = normalized_data[:, dim : dim + field.getOutputDim()]
         else:
             sub_data = normalized_data[:, :, dim : dim + field.getOutputDim()]
+
         sub_data = field.denormalize(sub_data)
-        if isinstance(field, ContinuousField):
-            sub_data = sub_data[:, 0] if is_session_key else sub_data[:, :, 0]
+        if field.name in word2vec_cols_names:
+            word2vec_type_col = None
+            for k in word2vec_dict_type_cols.keys():
```
Comment (saart) [inline netshare/generate/denormalize_fields.py]:
```suggestion
            for k in word2vec_dict_type_cols:
```
Just a syntax sugar :-)

File: netshare/generate/denormalize_fields.py (hunk 12 above)
Comment (saart) [reply netshare/generate/denormalize_fields.py]: Another benefit will be that the rest of your logic will use the `Field` object, and not `Any`, such as this function signature: `get_word2vec_type_col(word2vec_cols: List[Any])`

File: netshare/learn/utils/word2vec_embedding.py [hunk 14]
```diff
@@ -36,3 +41,90 @@ def word2vec_train(
...
+        type_dict = {}
+        index = 0
+
+        for ele in type_set:
+            type_ann.add_item(index, get_vector(model, str(ele), norm_option=True))
+            type_dict[index] = ele
+            index += 1
+        type_ann.build(n_trees)
+
+        dict_type_annDictPair[type] = (type_ann, type_dict)
```
Comment (saart) [inline netshare/learn/utils/word2vec_embedding.py]: The type of this structure is pretty complex. It is pretty hard to understand how to use it. e.g. when you wrote `dict_type_annDictPair[word2vec_type_col][0]` it is pretty hard to follow, and this is a source of bugs.
WDYT about using `namedtuple` here?
Here's a suggestion:
```
from typing import NamedTuple

class annoyTypeDescription(NamedTuple):
   annoy_type: int
   annoy_dict: Dict[int, ???]  # I really don't know
```
and then your return value will be: `Dict[str, annoyTypeDescription]`

File: netshare/generate/model_generate.py (hunk 7 above)
Comment (saart) [reply netshare/generate/model_generate.py:94]: If it works for you, then we can keep it like that, and I'll change it after the big merge !

File: netshare/generate/model_generate.py (hunk 6 above)
Comment (jinminhao) [reply netshare/generate/model_generate.py]: Yes, `word2vec_vecSize` is no longer usable.

File: netshare/generate/model_generate.py (hunk 7 above)
Comment (jinminhao) [reply netshare/generate/model_generate.py:94]: I agree. `List` or `OrderedDict` is definitely better. I also have this concern and when I'm iterating the `dict`, I find it out the order is correct. So I just plan to put it there and discuss it with you. I agree that using list is better, I will make the change on the code. Thanks!!

File: netshare/generate/denormalize_fields.py (hunk 11 above)
Comment (jinminhao) [reply netshare/generate/denormalize_fields.py]: Sure. I will also change a bit on `word2vec_train`

File: netshare/generate/denormalize_fields.py (hunk 12 above)
Comment (jinminhao) [reply netshare/generate/denormalize_fields.py]: Yes, I agree that adding word2vec flag into field is a great idea. Previously, I also had a chat with @sniperyyc on this stuff. I'm not doing that because the modification may include a lot of files. I'm bit afraid to have a large conflit with you on the branch `code-refactor` if you have some continue update on that. Since the work on `code-refactor` is almost done, I can do that!

File: netshare/learn/utils/word2vec_embedding.py (hunk 14 above)
Comment (jinminhao) [reply netshare/learn/utils/word2vec_embedding.py]: I think so. I can change it.

File: netshare/generate/model_generate.py (hunk 1 above)
Comment (jinminhao) [reply netshare/generate/model_generate.py:84]: No wories. It's because `apply_timestamp_generation` being called in `apply_distributed_chunk_logic` is after `build_fields` being called in `setup_cross_chunks_data`

File: netshare/generate/model_generate.py (hunk 7 above)
Comment (saart) [reply netshare/generate/model_generate.py:94]: Great, thank you 🙏 
If you prefer, you can leave it like that and I'll fix it in a followup PR

File: netshare/generate/denormalize_fields.py (hunk 12 above)
Comment (saart) [reply netshare/generate/denormalize_fields.py]: I don't have any work left in `code-refactor`, so don't worry about more conflicts :)

Comment (sniperyyc) [reply netshare/generate/denormalize_fields.py:62]: > 

Yes, I remembered our discussion. Adding a `word2vec` field is definitely a more generic/long-term solution to this.

Comment (sniperyyc) [reply netshare/generate/denormalize_fields.py]: > Yes, I agree that adding word2vec flag into field is a great idea. Previously, I also had a chat with @sniperyyc on this stuff. I'm not doing that because the modification may include a lot of files. I'm bit afraid to have a large conflit with you on the branch `code-refactor` if you have some continue update on that. Since the work on `code-refactor` is almost done, I can do that!

Yes, I remembered our discussion. Adding a `word2vec` field is definitely a more generic/long-term solution to this.

File: netshare/generate/model_generate.py (hunk 7 above)
Comment (jinminhao) [reply netshare/generate/model_generate.py:94]: Sure. I will leave it to you.

File: netshare/generate/model_generate.py (hunk 1 above)
Comment (saart) [reply netshare/generate/model_generate.py:84]: Ohhh I see, wow, so sorry for that :(
If you prefer, we can also move the logic, or change the place where we put the new fields (i.e. the line `new_timeseries_list.insert(0, time_col)`)...
Amazing that you spot it 👑

File: netshare/utils/field.py [hunk 15]
```diff
@@ -191,6 +206,53 @@ def getOutputDim(self) -> int:
         return 2 * self.num_bits  # type: ignore
 
 
+class Word2VecField(Field):
```
Comment (saart) [inline netshare/utils/field.py:218]: can you add typing here?
Maybe I'll add it as a constraint in mypy on my next PR :)

File: netshare/utils/word2vec_utils.py [hunk 16]
```diff
@@ -0,0 +1,9 @@
+import os
```
Comment (saart) [inline netshare/utils/word2vec_utils.py]: I think that this is just an empty file?

File: netshare/learn/setup_cross_chunks_data.py [hunk 17]
```diff
@@ -102,26 +107,39 @@ def get_word2vec_model(df: pd.DataFrame) -> Optional[Word2Vec]:
...
+    word2vec_train(
         df=df,
-        out_dir=get_word2vec_model_directory(),
         model_name=word2vec_config["word2vec"]["model_name"],
         word2vec_cols=word2vec_cols,
         word2vec_size=word2vec_config["word2vec"]["vec_size"],
         annoy_n_trees=word2vec_config["word2vec"]["annoy_n_trees"],
     )
-    return Word2Vec.load(word2vec_model_path)
+    word2vec_model_path = get_word2vec_model_path()
```
Comment (saart) [inline netshare/learn/setup_cross_chunks_data.py:128]: I thought that `build_annoy_dictionary_word2vec` will call this function itself, and not get it as an input,
But that's really not mandatory

File: netshare/learn/utils/word2vec_embedding.py [hunk 18]
```diff
@@ -35,4 +46,99 @@ def word2vec_train(
...
+
+    dict_type_annDictPair = {}
+
+    model = Word2Vec.load(model_path)
+    wv = model.wv
+
+    # type : [cols]
+    # ("ip": ["srcip", "dstip"])
+    # "word2vec_port": ["srcport", "dstport"]
+    # "word2vec_proto": ["proto"]
```
Comment (saart) [inline netshare/learn/utils/word2vec_embedding.py]: Can you please delete?

File: netshare/learn/utils/word2vec_embedding.py [hunk 19]
```diff
@@ -35,4 +46,99 @@ def word2vec_train(
...
+        type_set = set(
+            list(itertools.chain.from_iterable([list(df[col]) for col in cols]))
+        )
+        type_ann = AnnoyIndex(word2vec_size, "angular")
+        type_dict = {}
+        index = 0
+
+        for ele in type_set:
+            type_ann.add_item(index, get_vector(model, str(ele), norm_option=True))
+            type_dict[index] = ele
```
Comment (saart) [inline netshare/learn/utils/word2vec_embedding.py]:
```suggestion
        for index, ele in enumerate(type_set):
            type_ann.add_item(index, get_vector(model, str(ele), norm_option=True))
            type_dict[index] = ele
```
Syntax sugar :)

Comment (saart) [reply netshare/learn/utils/word2vec_embedding.py]: don't you prefer that `type_dict` will be a list?

File: netshare/learn/utils/word2vec_embedding.py [hunk 20]
```diff
@@ -35,4 +46,99 @@ def word2vec_train(
...
+        type_ann.save(get_annoyIndex_for_word2vec(type))
+        with open(get_annoy_type_dict_for_word2vec(), "w") as outfile:
+            json.dump(dict_type_annDictPair, outfile)
+
+    logger.info("Finish building Angular trees...")
+
+    return None
+
+
+def get_original_obj(ann, vector, dic):
```
Comment (saart) [inline netshare/learn/utils/word2vec_embedding.py]: Can you add types to all the functions in this file?

File: netshare/utils/field.py [hunk 21]
```diff
@@ -191,6 +206,53 @@ def getOutputDim(self) -> int:
...
+        return np.array(
+            [get_vector(embed_model, str(xi), norm_option=True) for xi in x]
+        )
+
+    def denormalize(self, norm_x):
+        dict_annDictPair = {}
+        with open(get_annoy_type_dict_for_word2vec(), "r") as readfile:
+            dict_annDictPair = json.load(readfile)
+        dict_annoyIndex = {}
+        for type in dict_annDictPair.keys():
```
Comment (saart) [inline netshare/utils/field.py]:
```suggestion
        for type in dict_annDictPair:
```

File: netshare/utils/field.py [hunk 22]
```diff
@@ -191,6 +206,53 @@ def getOutputDim(self) -> int:
...
+            type_ann = AnnoyIndex(self.word2vec_size, "angular")
+            type_ann.load(get_annoyIndex_for_word2vec(type))
+            dict_annoyIndex[type] = type_ann
+
+        for k in self.word2vec_dict_type_cols:
+            if self.name in self.word2vec_dict_type_cols[k]:
+                word2vec_type_col = k
+                break
+        if word2vec_type_col is None:
```
Comment (saart) [inline netshare/utils/field.py]: you might get here `AttributeError`. 
You probably would have found it if your wrote tests - but it's not mandatory, so it's your choice :)

Comment (saart) [reply netshare/utils/field.py]: btw, a pretty cool feature that does what you're trying to do here is `for...else`
Google it :)

File: netshare/learn/utils/word2vec_embedding.py (hunk 19 above)
Comment (jinminhao) [reply netshare/learn/utils/word2vec_embedding.py]: `type_dict`, is used to help to find out the nearest 'word' when denormalizing. Therefore, using dict for fast search for the correct 'word' may be important..

File: netshare/utils/word2vec_utils.py (hunk 16 above)
Comment (jinminhao) [reply netshare/utils/word2vec_utils.py]: Yes, sorry about that:)

File: netshare/learn/utils/word2vec_embedding.py (hunk 20 above)
Comment (jinminhao) [reply netshare/learn/utils/word2vec_embedding.py]: Sure

File: netshare/generate/denormalize_fields.py [hunk 23]
```diff
@@ -52,26 +66,34 @@ def write_to_csv(
...
+            session_data = session_key[i].tolist()
             # this if is here in parallel to the if in `reduce_samples`. It supports old flows.
-            if len(timeseries) == 1:
-                timeseries_data = timeseries[0][i].tolist()
-            else:
-                timeseries_data = [d[i][0] for d in timeseries]
-            writer.writerow(session_data + timeseries_data)
+            for j in range(data_gen_flag.shape[1]):
+                if data_gen_flag[i][j] == 1.0:
+                    timeseries_data = timeseries[i][j].tolist()
```
Comment (jinminhao) [inline netshare/generate/denormalize_fields.py]: When writing to csv, we need to rely on the gen flag to check if the session has already ended at a specific time step.

File: netshare/utils/field.py [hunk 24]
```diff
@@ -44,7 +59,14 @@ def normalize(self, x):
...
             )
         if self.log1p_norm:
+            # If set log1p_norm, we need to re-calculate the min_x and max_x after doing
+            # log1p. Since the denormalizing step would first do min-max denomalization
+            # and then expm1. Using stale min_x and max_x would cause the synthetic
+            # value larger than expected and is possible to have np.inf
             x = np.log1p(x)
+            self.min_x = np.log1p(self.min_x)
+            self.max_x = np.log1p(self.max_x)
+
```
Comment (jinminhao) [inline netshare/utils/field.py:70]: When activating log1p_norm, we need also to norm the min and max. Because in denoamlizing stage, the generated data will first be denomalized by min-max denoamlizagion, then expm1. Using a unnormalized min max may made the result larger then expected and sometimes may cause it become inf.

File: netshare/generate/denormalize_fields.py [hunk 25]
```diff
@@ -32,14 +36,24 @@ def _denormalize_by_fields_list(
...
             sub_data = normalized_data[:, dim : dim + field.getOutputDim()]
         else:
             sub_data = normalized_data[:, :, dim : dim + field.getOutputDim()]
+
         sub_data = field.denormalize(sub_data)
-        if isinstance(field, ContinuousField):
-            sub_data = sub_data[:, 0] if is_session_key else sub_data[:, :, 0]
+
+        # For session key, if shape looks like (n, ), change it to (n, 1) for consistency
+        if is_session_key == True and len(sub_data.shape) == 1:
```
Comment (saart) [inline netshare/generate/denormalize_fields.py]:
```suggestion
        if is_session_key and len(sub_data.shape) == 1:
```
Search for "greeting" in https://peps.python.org/pep-0008/

File: netshare/generate/denormalize_fields.py [hunk 26]
```diff
@@ -32,14 +36,24 @@ def _denormalize_by_fields_list(
...
+            sub_data = np.expand_dims(sub_data, axis=1)
+        # For timeseries, if shape looks like (i, j), change it to (i, j, 1) for consistency
+        if is_session_key == False and len(sub_data.shape) == 2:
```
Comment (saart) [inline netshare/generate/denormalize_fields.py]:
```suggestion
        if not is_session_key and len(sub_data.shape) == 2:
```

File: netshare/generate/denormalize_fields.py (hunk 25 above)
Comment (saart) [reply netshare/generate/denormalize_fields.py]: (absolutely nit-picking, you can ignore if you wish)

File: netshare/generate/denormalize_fields.py [hunk 27]
```diff
@@ -52,26 +66,34 @@ def write_to_csv(
...
+
     with open(csv_path, "w") as f:
         writer = csv.writer(f)
         session_titles = _get_fields_names(session_key_fields)
         timeseries_titles = _get_fields_names(timeseries_fields)
         writer.writerow(session_titles + timeseries_titles)
 
-        for i in range(session_key[0].shape[0]):
-            session_data = [d[i] for d in session_key]
+        for i in range(data_gen_flag.shape[0]):
```
Comment (saart) [inline netshare/generate/denormalize_fields.py]:
```suggestion
        for data_gen, session_data, timeserie in zip(data_gen_flag, session_key, timeseries):
```
And then you can forget about the index `i`.
(Please check it out - I'm not sure that it will work with np.arrays, but if it doesn't - it helps for readability)

File: netshare/learn/setup_cross_chunks_data.py [hunk 28]
```diff
@@ -84,13 +88,15 @@ def get_global_max_flow_len(df_chunks: List[pd.DataFrame]) -> int:
     return max(max_flow_lens)
 
 
-def get_word2vec_model(df: pd.DataFrame) -> Optional[Word2Vec]:
+def get_word2vec_model(
+    df: pd.DataFrame,
+) -> Union[Tuple[Word2Vec, List[str]], Tuple[None, None]]:
```
Comment (saart) [inline netshare/learn/setup_cross_chunks_data.py]:
```suggestion
) -> Tuple[Optional[Word2Vec], Optional[List[str]]]:
```
maybe?

File: netshare/learn/utils/word2vec_embedding.py [hunk 29]
```diff
@@ -35,4 +46,99 @@ def word2vec_train(
...
+
+
+def get_word2vec_type_col(word2vec_cols: List[Any]) -> Dict[str, List[str]]:
+    dict_type_cols: Dict[str, List[str]] = {}
+    for col in word2vec_cols:
+        type = col.encoding
+        if type not in dict_type_cols:
+            dict_type_cols[type] = []
+        dict_type_cols[type].append(col.column)
+    return dict_type_cols
```
Comment (saart) [inline netshare/learn/utils/word2vec_embedding.py]:
```suggestion
    dict_type_cols: Dict[str, List[str]] = defaultdict(list)
    for col in word2vec_cols:
        dict_type_cols[col.encoding].append(col.column)
    return dict_type_cols
```
`from collections import defaultdict` This is one of the best utilities :)

File: netshare/learn/utils/word2vec_embedding.py [hunk 30]
```diff
@@ -35,4 +46,99 @@ def word2vec_train(
...
+    n_trees: int,
+) -> None:
+
+    dict_encoding_type_vs_idx_ele_dict_pair = {}
+
+    model = Word2Vec.load(model_path)
+    wv = model.wv
+
+    # encoding type and column names may not be necessarily the same. eg. for a pcap/netflow dataset
+    # encoding type = word2vec_port -> column: ["srcport", "dstport"]
```
Comment (saart) [inline netshare/learn/utils/word2vec_embedding.py:74]: so the type is structured from two different entities? 
`word2vec` which tells us the type, and `port`/`proto` which tells us the "group of words" that it should be included in?
Why isn't that another key in the field's configuration, and just maintain it as a member of `Word2VecField`?

Comment (saart) [reply netshare/learn/utils/word2vec_embedding.py:74]: And then, you can use `word2vec_fields: List[Field]` instead of `word2vec_cols: List[Any]` everywhere.

Comment (saart) [reply netshare/learn/utils/word2vec_embedding.py:74]: But: Maybe this is not in the scope of this PR, and we should do this change later.
It's your choice :)

File: netshare/learn/utils/word2vec_embedding.py [hunk 31]
```diff
@@ -35,4 +46,99 @@ def word2vec_train(
...
+    # encoding type = word2vec_proto -> column: ["proto"]
+    # srcport and dstport are two fields for a dataset, but from a word2vec perspective, they are both ports
+    dict_encoding_type_vs_cols = get_word2vec_type_col(word2vec_cols)
+    logger.debug(f"word2vec type columns are: {dict_encoding_type_vs_cols}")
+
+    for encoding_type, cols in dict_encoding_type_vs_cols.items():
+        ele_set = set(
+            list(itertools.chain.from_iterable([list(df[col]) for col in cols]))
```
Comment (saart) [inline netshare/learn/utils/word2vec_embedding.py]:
```suggestion
          itertools.chain.from_iterable([list(df[col]) for col in cols])
```
Much faster :-)

File: netshare/learn/utils/word2vec_embedding.py [hunk 32]
```diff
@@ -35,4 +46,99 @@ def word2vec_train(
...
+def get_original_objs(
+    ann: AnnoyIndex, vectors: np.ndarray, dic: Dict[int, Any]
+) -> List[Any]:
+    res = []
+    for vector in vectors:
+        obj_list = ann.get_nns_by_vector(
+            vector, 1, search_k=-1, include_distances=False
+        )
+        res.append(dic[obj_list[0]])
+    return res
```
Comment (saart) [inline netshare/learn/utils/word2vec_embedding.py]:
```suggestion
    return [get_original_obj(ann, vector, dic) for vector in vectors]
```
DRY
right?

File: netshare/learn/utils/word2vec_embedding.py [hunk 33]
```diff
@@ -35,4 +46,99 @@ def word2vec_train(
...
+
+def get_vector(model: Word2Vec, word: str, norm_option: bool = False) -> np.ndarray:
+    all_words_str = list(model.wv.vocab.keys())
+
+    # Privacy-related
+    # If word not in the vocabulary, replace with nearest neighbor
+    # Suppose that protocol is covered
+    #   while very few port numbers are out of range
+    if word not in all_words_str:
+        # print(f"{word} not in dict")
```
Comment (saart) [inline netshare/learn/utils/word2vec_embedding.py]: delete

File: netshare/utils/field.py (hunk 24 above)
Comment (saart) [reply netshare/utils/field.py:70]: That's a good catch!
Can you please move this code to the __init__?
`normalize` should be a pure function: I.e. we don't expect it to change any internal value while operating. I.e. someone can call it multiple times and that's fine. Here, you're changing members inside `self`

File: netshare/utils/field.py [hunk 34]
```diff
@@ -191,6 +214,57 @@ def getOutputDim(self) -> int:
...
+        for k in self.dict_encoding_type_vs_cols:
+            if self.name in self.dict_encoding_type_vs_cols[k]:
+                encoding_type = k
+                break
+        else:
+            raise ValueError("Cannot find the word2vec key!")
+        # When constructing the dict_annDictPair, dict_annDictPair[encoding_type]: dict{}
+        # is a k v pair where k is an integer. After using Json.save() then Json.load(), it
+        # will change k into str. Therefore, when we are using dict_annDictPair[encoding_type]
+        # the k should be casted to int.
```
Comment (saart) [inline netshare/utils/field.py:250]: sounds like an awful bug 😂 
Well done!!!
Here's a relevant link: https://stackoverflow.com/questions/1450957/pythons-json-module-converts-int-dictionary-keys-to-strings

**Review by saart — APPROVED** [review]
LGTM!
Well done!!

File: netshare/generate/denormalize_fields.py (hunk 25 above)
Comment (jinminhao) [reply netshare/generate/denormalize_fields.py]: It's definitely more elegent. Thanks!!

File: netshare/learn/utils/word2vec_embedding.py (hunk 30 above)
Comment (jinminhao) [reply netshare/learn/utils/word2vec_embedding.py:74]: Great idea! i think so. Let me mark it and put it into another PR

**Comment by jinminhao** [comment]
> LGTM! Well done!!

Thank you so much for your review and comments!!! I just gave a final update and will merge it.
